File: frontend/lib/api/analyze.ts
import { apiBase, fetchJson, onUnauthorizedResponse, withAuth } from "./api";

export type Citation = { document_id: number; excerpt: string; source_path?: string | null };

export type AnalyzeResponse = {
  answer: string;
  key_points: string[];
  citations: Citation[];
};

export type AnalyzeBody = {
  question: string;
  reindex_sources: boolean;
  top_k?: number;
  /** 既定 true。false で DB 履歴に残さない */
  save_question_history?: boolean;
};

export async function postAnalyze(body: AnalyzeBody): Promise<AnalyzeResponse> {
  return fetchJson<AnalyzeResponse>(`${apiBase()}/api/analyze`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(body),
  });
}

export type StreamEvent =
  | { type: "token"; content: string }
  | { type: "done"; key_points: string[]; citations: Citation[] }
  | { type: "error"; message: string };

export async function* postAnalyzeStream(
  body: AnalyzeBody,
): AsyncGenerator<StreamEvent> {
  const resp = await fetch(
    `${apiBase()}/api/analyze/stream`,
    withAuth({
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(body),
    }),
  );

  onUnauthorizedResponse(resp);

  if (!resp.ok) {
    const text = await resp.text();
    let detail = text;
    try {
      const j = JSON.parse(text) as { detail?: unknown };
      if (j.detail != null) detail = JSON.stringify(j.detail);
    } catch { /* use text */ }
    throw new Error(`${resp.status} ${detail}`);
  }

  const reader = resp.body?.getReader();
  if (!reader) throw new Error("No response body");

  const decoder = new TextDecoder();
  let buffer = "";

  try {
    while (true) {
      const { done, value } = await reader.read();
      if (done) break;
      buffer += decoder.decode(value, { stream: true });

      const lines = buffer.split("\n");
      buffer = lines.pop() ?? "";

      for (const line of lines) {
        if (!line.startsWith("data: ")) continue;
        try {
          yield JSON.parse(line.slice(6)) as StreamEvent;
        } catch { /* malformed line, skip */ }
      }
    }
  } finally {
    reader.releaseLock();
  }
}

File: backend/app/schemas/analyze.py
from pydantic import BaseModel, Field


class Citation(BaseModel):
    """回答の根拠となったドキュメント断片。"""

    document_id: int = Field(description="documents テーブルの主キー")
    excerpt: str = Field(
        max_length=2000,
        description="引用した本文の抜粋（数百文字以内推奨）",
    )


class AnalyzeRequest(BaseModel):
    question: str = Field(min_length=1, max_length=8000)
    reindex_sources: bool = Field(
        default=False,
        description="true のとき DATA_DIR を再取り込み（documents / raw_data を置き換え）",
    )
    top_k: int | None = Field(
        default=None,
        ge=1,
        le=20,
        description="ベクトル検索の件数。省略時は環境変数 RAG_TOP_K",
    )
    save_question_history: bool = Field(
        default=True,
        description="true のとき成功応答を question_history に保存する",
    )


class AnalyzeResponse(BaseModel):
    answer: str = Field(description="質問に対する日本語または入力言語での要約回答")
    key_points: list[str] = Field(
        description="箇条書きの重要ポイント（短い文のリスト）",
    )
    citations: list[Citation] = Field(
        description="コンテキストに用いた documents の引用",
    )
